File: nodejs/dvh.js
const crypto = require('crypto');
 
function createHmac (authObject, secretKey) {
   try {
     let dvhString =  Buffer.from(JSON.stringify(authObject)).toString('base64');
     return crypto.createHmac(`SHA512`, secretKey).update(dvhString).digest('hex');
   } catch (error) {
     throw error;
   }
 };

let request = {
 "amount": 100.5,
 "apiKey": "3568f8c7-3f33-49dc-bbc9-9362c130f7c8",
 "bankCode": "GIBL",
 "currency": "NPR",
 "dateOfRequest": "2020-09-10",
 "dvh": "01f0bb5021822af7dc911e2d2a1684930fd5887f38da909d33b37641d14baa842c1c103e30b31a135e4615bcbe0d1a7b26c2abf5dda8c4c3680104a0e88006bf",
 "referenceId": "Drz9699731138518"
}

 let { dvh, metaData, context, ...validationObject } = request;   
 let secretKey ='3568f8c73f3349dcbbc99362c130f7c8'   

request.dvh = createHmac(validationObject,secretKey )

 console.log(JSON.stringify(request))


 

File: c#/dvh.cs
using System;
using System.Text;
using System.Security.Cryptography;
using Newtonsoft.Json ;
public class TokenRequestModel
{
	
    public decimal amount { get; set; }
	public string apiKey { get; set; }
	public string bankCode { get; set; }
	public string currency { get; set; }
	public string dateOfRequest { get; set; }
	public string referenceId { get; set; }
		
}

public class Program
{
	public static void Main()
	{
		var account = new TokenRequestModel() { amount = 100.5M , apiKey = "3568f8c7-3f33-49dc-bbc9-9362c130f7c8" , bankCode = "GIBL" ,  currency= "NPR" ,dateOfRequest = "2020-09-10" , referenceId = "Drz9699731138518"  };
        string json = JsonConvert.SerializeObject(account);
		string ipnSecret = "3568f8c73f3349dcbbc99362c130f7c8";
		Console.WriteLine("json : "+ json);
		// string json = "{\"amount\":100.5,\"apiKey\":\"3568f8c7-3f33-49dc-bbc9-9362c130f7c8\",\"bankCode\":\"GIBL\",\"currency\":\"NPR\",\"dateOfRequest\":\"2020-09-10\",\"referenceId\":\"Drz9699731138518\"}";
		string base64EncodedExternalAccount = Convert.ToBase64String(Encoding.UTF8.GetBytes(json));
		string hmac = calcHMAC(base64EncodedExternalAccount , ipnSecret);
		Console.WriteLine("hmac : "+ hmac);
		
	}
	public static string calcHMAC(string parms, string ipnSecret)
	{
		byte[] keyBytes = Encoding.UTF8.GetBytes(ipnSecret);
		byte[] postBytes = Encoding.UTF8.GetBytes(parms);
		var hmacsha512 = new HMACSHA512(keyBytes);
		string hmac = BitConverter.ToString(hmacsha512.ComputeHash(postBytes)).Replace("-", "").ToLowerInvariant();
		return hmac;
	}
	
}